{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "positive-boutique",
   "metadata": {},
   "source": [
    "## HW4. RNA and Clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "occasional-student",
   "metadata": {},
   "source": [
    "### In this homework, we will analyze breast cancer RNA-seq data generated by TCGA in order to re-discover the clinically used BRCA subtypes defined by the PAM50 assay. We will be making use of the following files\n",
    "\n",
    "- EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp_BRCAsubset.tsv.gz. This file contains a gene x sample matrix of normalized RNA expression quantifications subset to BRCA samples download from the GDC website https://gdc.cancer.gov/about-data/publications/pancanatlas. See here for a description of their quantification pipeline: https://gdc.cancer.gov/about-data/data-harmonization-and-generation/genomic-data-harmonization/high-level-data-generation/rna-seq-quantification.\n",
    "- BRCA_clinical.txt. This file gives us clinical information about the patients from which tumors were samples, as reported in the following TCGA paper https://www.cell.com/cancer-cell/pdfExtended/S1535-6108(18)30119-3. Of particular interest is the column \"BRCA_Subtype_PAM50\" which categorizes BRCA into the following subtypes with distinct expression, clinical properties, and cell of origin\n",
    "\n",
    "  - Basal\n",
    "  - Her2\n",
    "  - LumA\n",
    "  - LumB\n",
    "  - Normal-like\n",
    "  \n",
    "Using PCA and the clustering techniques in class, we will characterize the main sources of variation in this dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "black-challenge",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "heard-electric",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = pd.read_csv('EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp_BRCAsubset.tsv.gz',sep='\\t',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "protected-concentration",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-3C-AAAU-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-3C-AALI-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-3C-AALJ-01A-31R-A41B-07</th>\n",
       "      <th>TCGA-3C-AALK-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-4H-AAAK-01A-12R-A41B-07</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>197.090</td>\n",
       "      <td>237.3840</td>\n",
       "      <td>423.2370</td>\n",
       "      <td>191.0180</td>\n",
       "      <td>268.8810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9066</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.4255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2BP1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3.8298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2LD1</th>\n",
       "      <td>102.963</td>\n",
       "      <td>70.8646</td>\n",
       "      <td>161.2600</td>\n",
       "      <td>62.5072</td>\n",
       "      <td>154.3700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>5798.370</td>\n",
       "      <td>7571.9800</td>\n",
       "      <td>8840.4000</td>\n",
       "      <td>10960.2000</td>\n",
       "      <td>9585.4400</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         TCGA-3C-AAAU-01A-11R-A41B-07  TCGA-3C-AALI-01A-11R-A41B-07  \\\n",
       "gene_id                                                               \n",
       "A1BG                          197.090                      237.3840   \n",
       "A1CF                            0.000                        0.0000   \n",
       "A2BP1                           0.000                        0.0000   \n",
       "A2LD1                         102.963                       70.8646   \n",
       "A2M                          5798.370                     7571.9800   \n",
       "\n",
       "         TCGA-3C-AALJ-01A-31R-A41B-07  TCGA-3C-AALK-01A-11R-A41B-07  \\\n",
       "gene_id                                                               \n",
       "A1BG                         423.2370                      191.0180   \n",
       "A1CF                           0.9066                        0.0000   \n",
       "A2BP1                          0.0000                        0.0000   \n",
       "A2LD1                        161.2600                       62.5072   \n",
       "A2M                         8840.4000                    10960.2000   \n",
       "\n",
       "         TCGA-4H-AAAK-01A-12R-A41B-07  \n",
       "gene_id                                \n",
       "A1BG                         268.8810  \n",
       "A1CF                           0.4255  \n",
       "A2BP1                          3.8298  \n",
       "A2LD1                        154.3700  \n",
       "A2M                         9585.4400  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.iloc[0:5,0:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "neural-inside",
   "metadata": {},
   "source": [
    "# Question 1. Data Exploration"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "authorized-madonna",
   "metadata": {},
   "source": [
    "## (1a) Calculate the average expression of genes across the cohort and make a histogram (Hint: for better visualization you may want to apply a simple transform). Based on the observed distribution, choose a cutoff and filter \"non-expressed\" genes in the peak near zero expression. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "specialized-france",
   "metadata": {},
   "source": [
    "## (1b) Using the PAM50 subtype annotations provided in the clinical data, visualize the expression of the following genes stratified by subtype to confirm they are indeed marker genes for the stated subtype\n",
    "\n",
    "- ERBB2 (HER2)\n",
    "- ESR1 (LumA/B)\n",
    "- KRT5 (Basal)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "referenced-heavy",
   "metadata": {},
   "source": [
    "# Question 2. Dimensionality reduction\n",
    "\n",
    "## (2a) Run PCA on the (log(x+1)-transformed) data. You can use an implementation such as sklearn, but explain which parts of the result object are the eigenvectors and values of the data's covariance matrix we described in class. Create a scatterplot of the data in PC-space for the first two dimensions colored by PAM50 classification. Do you see a separation of the subtypes?\n",
    "\n",
    "## (2b) Explore the gene loadings (i.e. weightings in the eigenvector) of the first two components. For each, look up the few highest and lowest scored genes on a site like genecards.org and hypothesize the biological signal underlying the variation in each dimension. Do they have to do with tumor subtype? Or is there some confounding source of variation?\n",
    "\n",
    "## (2c) Make a plot of the variance explained by each for the first 50 principal components. Using the \"elbow method\", how many PCs is reasonable to take for downstream analysis?\n",
    "\n",
    "## (2d) Run one of the non-linear embedding techniques discussed in class (e.g. t-SNE) on your PCA-transformed data, using the number of PCs you determined in 2c. Plot the data colored by PAM50. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "personal-relay",
   "metadata": {},
   "source": [
    "# Question 3. Clustering\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "incredible-combining",
   "metadata": {},
   "source": [
    "## (3a) Cluster the data using one of the techniques discussed in class (k-means, graph-based, etc) using your PCA-transformed data. Is there a number of clusters/resolution that reflects the PAM50 subtypes? Create a visualization to demonstrate the extent of the agreement."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "swedish-smile",
   "metadata": {},
   "source": [
    "## (3b) Find and plot marker genes for each cluster: Using a t-test, test for each gene whether expression is significantly different between samples that are inside vs outside of the cluster. If you have K clusters and G genes, you should end up with $K*G$ p-values. Find the top 10 marker genes for each cluster (10xK total genes) and plot a heatmap of the expression of these across all the samples. Group samples by their cluster, so you should see blocks of the heat map light up (e.g. the first 10 genes should light up in samples of the first cluster and be dark in the rest, the next 10 high in the second cluster, etc). If you have clusters not corresponding to the PAM50 subtypes, what signal do you think drives them?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "tropical-malaysia",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
